### PR TITLE
Add SCons generator

### PR DESCRIPTION
**Pull request by pepsiman**
Suggested usage in SConstruct file:

conan = SConscript('SConscript-conan')

completeEnv = Environment()
completeEnv.MergeFlags(conan['conan'])
completeEnv.Program("foo", "foo.c")

partialEnv = Environment()
partialEnv.MergeFlags(conan['zlib'])
partialEnv.MergeFlags(conan['poco'])
partialEnv.Program("bar", "bar.c")

### CODE REVIEW

File: conans/client/generators/scons.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from conans.model import Generator
+
+class SConsGenerator(Generator):
+    @property
+    def filename(self):
+        return "SConscript-conan"
```
Comment (memsharded) [inline conans/client/generators/scons.py]: All the genereators create all lowercase files, and prefer underscore. Is it possible to change the name, or is it somehow dictated by Scons? Maybe something like conanbuildinfo.scons?

**Comment by memsharded** [comment]
I like the idea of having SCons generator, and the implementation is good, just a couple of things to consider:

- If you think the generator can be quickly evolved, will need fixes, etc., maybe you want to consider this: http://docs.conan.io/en/latest/howtos/dyn_generators.html (I am not saying to do it, I think it can be a built-in generator)
- If the suggested usage approach seems to have a large reusable part (all the Appends), maybe just create a function inside the generator to do that and simplify usage?

Thanks for your contribution!

File: conans/client/generators/scons.py (hunk 1 above)
Comment (pepsiman) [inline conans/client/generators/scons.py]: The man page says "By convention, these subsidiary files are named SConscript, although any name may be used".
It doesn't really matter.

**Comment by pepsiman** [comment]
As SCons configuration files are python scripts, there is a lot of flexibility.
So I don't want to restrict how conan is used with it, but avoiding boilerplate would be good too.
I want to avoid adding the list of all dependent libs to the SCons environment as not all programs link against all libs.
It seemed simplest to just provide the data and let the project decide what to do with it.

I'm still trying to work out how to integrate conan into my project.
I can currently build both debug and release with a single invocation of scons.
I think conan only sets up paths for one build type at a time.

File: conans/client/generators/scons.py (hunk 1 above)
Comment (memsharded) [inline conans/client/generators/scons.py]: Perfect, then leave it as is, following the SCons convention, thanks!

**Comment by memsharded** [comment]
I agree, the consumer should not be imposed anything. I wouldn't impose the use of the code, but simple make it available, not sure about the import mechanism of SCons, but something like what it is done with cmake:

In the generated file:

```python

# variables here

def conan_basic_setup(env):
    env.Append(CPPPATH = env['CONAN_CPPPATH'])
    env.Append(LIBPATH = env['CONAN_LIBPATH'])
    env.Append(CPPDEFINES = env['CONAN_CPPDEFINES'])
    env.Append(CPPFLAGS = env['CONAN_CPPFLAGS'])
    env.Append(CCFLAGS = env['CONAN_CCFLAGS'])
    env.Append(SHLINKFLAGS = env['CONAN_SHLINKFLAGS'])
    env.Append(LINKFLAGS = env['CONAN_LINKFLAGS'])
```

Then, when you use it, you can do it yourself, or call it to avoid boilerplate (or even forgetting some variable):

```python
SConscript('SConscript-conan')
Import('conan_env')

env = conan_env
conan_basic_setup(env)

env.Program("foo", "foo.c", LIBS = env['ZLIB_LIBS'] + env['POCO_LIBS'])
env.Program("bar", "bar.c", LIBS = env['CONAN_LIBS'])
```

**Comment by memsharded** [comment]
The issue with both debug/release builds:

You might be interested in reading https://github.com/conan-io/conan/issues/330

In summary:

- Conan design promotes to have different packages for debug and release (like apt-get dbg packages), for most use cases it is the best approach
- When you install something, you just get what you want, if you don't need debug artifacts, they will not be retrieved and installed. Consider that debug artifacts can be huge in many cases.
- Why not also bundling 32/64 bits artifacts in the same package? Now we will have x4 the artifacts.
- Also, if the package allows shared/static libraries, why not bundle them in the same package? Now, x8 artifacts, for everything. You will get many artifacts you don't need, wasting bandwidth and disk space, and what is worse, if you want to build from sources, it will be 8 times slower.
- Switching from debug to release (or any other configuration), is extremely simple: just type something like: ``$ conan install -s build_type=Debug``. If you switch often, this will be in your history: 3 keystrokes. And the command runs fast enough, as everything is cached.

Said that, there might be the possibility to do this:
- Remove the ``build_type`` setting from the conanfile.py
- When you ``package()`` your artifacts, package them into ``libs/debug`` and ``libs/release``.
- In ``package_info()`` add all to the ``libs`` field: ``self.cpp_info.libs = ["debug/mylib1", "release/mylib1"...]``
- In the consumer code, split the ``libs`` according to their prefix.

This is something we might have a look again for next release, but it might be too complex for something that is elegantly solved by 3 keystrokes. I always think of it as "git branching", which I do many times. If I switch from Release -> Debug, my dependencies have to be installed for Debug, so I tell conan with ``conan install -s build_type=Debug``.

**Comment by pepsiman** [comment]
SCons does variant builds using a folder per configuration.
I'm experimenting with using subprocess.call() from an scons config file to run conan inside that folder with the appropriate settings.

**Comment by pepsiman** [comment]
I think I can make this work:
```
conan = SConscript('SConscript-conan')

completeEnv = Environment()
completeEnv.MergeFlags(conan['conan'])
completeEnv.Program("foo", "foo.c")

partialEnv = Environment()
partialEnv.MergeFlags(conan['zlib'])
partialEnv.Program("bar", "bar.c")
```

File: conans/client/generators/scons.py (hunk 1 above)
Comment (pepsiman) [inline conans/client/generators/scons.py]: Hyphen or underscore isn't part of the convention, do you want that changed?

Comment (memsharded) [inline conans/client/generators/scons.py]: Yep, maybe better underscore

**Comment by memsharded** [comment]
Sounds good. Let's do it for next release then. Everything good on your side? Ready to be merged? Can you already use it properly for your projects? I usually try to test with a diamond of dependencies, something like A->B, A->C, B->D, C->D, to ensure that variables and everything are good. Thanks!

**Comment by pepsiman** [comment]
I saw the diamond tests in the Conan source.
I'd like to have a test to verify that the file has serialised the python lists and strings correctly, but I think the content of those lists is tested already.
I'll change the - to _ now and add tests when I have time. When is the release?
I've got a lot of packages to create before I can use it properly, but I have replaced my use of zlib with a Conan installed zlib as a proof of concept.

**Comment by memsharded** [comment]
Ok, perfect.

The release date is not scheduled yet, we are just selecting features and scope, still have to estimate. No less than two weeks, so plenty of time to add tests for it. We can do whatever you prefer: we can merge this right now, and have new PR for adding tests and possible improvements, or just keep working on this PR.

Let's share it, trying to get more feedback too. Thanks!

**Comment by pepsiman** [comment]
I added a test based on cmake_test.py

File: conans/client/generators/scons.py (hunk 1 above)
Comment (bdbaddog) [inline conans/client/generators/scons.py]: Is the intent that the file would be picked up by scons automatically? if so then the following from the manpage applies "

> By default, scons searches for a file named SConstruct, Sconstruct, or sconstruct (in that order) in the current directory and reads its configuration from the first file found. An alternate file name may be specified via the -f option.

**Comment by memsharded** [comment]
@bdbaddog One design guideline for conan is trying to be explicit on the user side. E.g. in the case of Cmake, conan generates a ``conanbuildinfo.cmake``. Including it does nothing, and the user has to explicitly call ``conan_basic_setup()`` or any of the other macros to actually set build flags. I like the approach of having a ``SCons_conan`` file that has to be explicitly imported and used. But yours was a valid point, so thanks very much for your feedback.

@pepsiman BTW, it is a common pattern that build folders might be different from project folder, like out-of-source builds. In cmake we use the idiom ``include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)`` which points to the build directory always. Generated files are different per configuration, and some users store builds (and generated files like SCons-conan file) in different folders, like ``Debug`` or ``Release64``

**Comment by bdbaddog** [comment]
@memsharded does conan create the build folders which these generated SCons files should live in and/or specify a directory tree structure for building variants?

**Comment by pepsiman** [comment]
SCons supports having multiple build directories, but they're all within the project directory.
The `SConstruct` at the root of the project specifies the build directories.
For my project, scons runs `conan install -f ../conanfile.py -s build_type=Debug ...` inside each build directory to generate the appropriate configuration, loads that configuration and merges the flags into the `Environment`.
The only difference between this usage and the suggested usage example is the `subprocess.call()` to run conan.

@bdbaddog conan only deals with one variant at a time.  It writes its configuration to files in the current directory.  The packaged libraries are kept separate from the project and build directories.

**Comment by memsharded** [comment]
@bdbaddog @pepsiman Actually, many users use multiple configurations, the typical workflow is:

```bash
$ mkdir debug32 && cd debug32
$ conan install .. -s build_type=Debug -s arch=x86 ....
#build, with conan build, cmake....
$ cd ..
$ mkdir release64 && cd release64
$ conan install .. -s build_type=Release -s arch=x86_64 ...
# build ....
# change code
# build
$ cd ../debug32
# build
```

As you can see the folder names are not automatically generated, they do not follow any pattern, just defined by the user, but the flow is simple and convenient to not having to regenerate/rebuild your whole project every configuration switch.

**Comment by bdbaddog** [comment]
@pepsiman Re "SCons supports having multiple build directories, but they're all within the project directory."

Not strictly true. The variant dirs can be anywhere.  The only gotcha is to remember that scons only builds from the current directory down by default. You'd need to add the variant dirs (if they not under the top level directory, or current working dir) as Default()..

**Comment by memsharded** [comment]
Hi!

@pepsiman , please we'd need that you sign the CLA, it appears as not signed

We are preparing conan 0.18, and would like to merge this. I think it can be merged, added to the docs, and hopefully we would get further feedback. Do you think that it should be put on hold @bdbaddog ? Any more specific advice of something that should be fixed before merging it?

About the multi-configuration builds, we are adding for this release support for CMake multi configuration (Debug/Release) projects, like that of VisualStudio and XCode. Besides the cmake complexities, it was not difficult, so if SCons can handle that, it should be doable. Please check: https://github.com/conan-io/conan/pull/804, and ask if you need some help.

Thanks very much!

**Comment by pepsiman** [comment]
That's weird; I remember signing it.
I'll have a look at #804.

**Comment by memsharded** [comment]
Great, thanks, I'll wait for your feedback about it, then.

**Comment by pepsiman** [comment]
Without #799, doing something like #804 would only mean changing the generated filename to include the build type.  The project's SCons files would have to decide which generated file to use in each variant.
SCons does not define standard build types.
With #799, a single file could contain all the data.  Similar logic would still be needed in the project.
I think I'll stick to running conan install inside each variant dir for now.

**Comment by memsharded** [comment]
Ok, good. If later we want to have another scons-multi generator, similar to the cmake_multi one, then we can make it too. Actually, I think the approach with one build folder per variant is better, more standard in all systems too. So mostly agree with your decision, merged, it will be available in next conan 0.18. Thanks for your contribution :)

**Pull request merged**